qiime2/core/testing/pipeline.py

```diff
@@ -195,6 +195,48 @@ def internal_fail_pipeline(ctx, ints1, ints2, int1, string, fail=False):
     return ints1_ret, ints2_ret, int1_ret
 
 
+def de_facto_list_pipeline(ctx, kwarg=False, non_proxies=False):
+    returns_int = ctx.get_action('dummy_plugin', 'returns_int')
+    list_of_ints = ctx.get_action('dummy_plugin', 'list_of_ints')
+    num_ints = 3
+
+    ints = []
+    for i in range(num_ints):
+        ints_ret, = returns_int(i)
+        ints.append(ints_ret)
+
+    if non_proxies:
+        ints.append(ctx.make_artifact(SingleInt, num_ints + 1))
+
+    if kwarg:
+        ret, = list_of_ints(ints=ints)
+    else:
+        ret, = list_of_ints(ints)
+
+    return ret
+
+
+def de_facto_dict_pipeline(ctx, kwarg=False, non_proxies=False):
+    returns_int = ctx.get_action('dummy_plugin', 'returns_int')
+    dict_of_ints = ctx.get_action('dummy_plugin', 'dict_of_ints')
+    num_ints = 3
+
+    ints = {}
+    for i in range(num_ints):
+        ints_ret, = returns_int(i)
+        ints[i + 1] = ints_ret
+
+    if non_proxies:
+        ints[num_ints + 2] = ctx.make_artifact(SingleInt, num_ints + 1)
```

Comment:
Same question as above, but also why you're adding 2 to this index

Reply:
Similar deal here to the above, but I am going to leave a +1 (changed from +2) in the key so the dict is {'1':0, '2':1, '3':2}. Early on in writing this stuff I tried to avoid having the keys in one of these dicts be monotonically increasing from 0. I avoided that because if things went wrong somehow and the dict turned into a list then was re-keyed it would be given keys that were monotonically increasing from 0. Any other set of keys in the output lets me know that it retained its keys the whole way through.